flybywiresim/a32nx
# feat(atsu): add PDC/CPDLC communication via Hoppie

## Summary of Changes
This PR adds PDC/CPDLC communication via the Hoppie.nl-interface to the A32NX.
It uses the MCDU and DCDU for the communication with the controllers.
The configuration is handled via the Settings-page of the EFB.

The CPDLC ring tone and the ATC button next to the master warning are added.

It integrates the full PDC communication and simple CPDLC instructions.
Instructions that require generated or modified answers or requests are out-of-scope for this changes.

Implemented ATC instructions:
- Logon/Logoff
- Handoff
- Station identifier
- Climb FLxxx, Fly heading xxx, Reduce/Increase speed xxx
- PDC

Introduced an ATSU manager with message formats for all AOC messages and parts of the ATC messages.
Rewrote ATSU/AOC pages to use a central ATSU manager and its types.
Introduced a central datalink manager to synchronize the messages and simulate single channel communication via VHF3.
Rewrote ATSU/ATC pages to make use of the ATSU manager.

The datalink transmission time supports three different types:
- Instant: Transmission of data takes up to two seconds
- Fast: Transmission of data takes up to twenty seconds
- Real: Transmission of data takes up to 120 seconds (simulate CPDLC 240 roundtrip-time)

This feature can only be used if the pilot flies on VATSIM or IVAO.

Deactivated all additional pages of "ATC MENU" that are currently not supported by the datalink.
Will be reactivated in later PRs.

**This PR depends on the extension of the NXApi and the corresponding PR: https://github.com/flybywiresim/api/pull/119**
## Screenshots (if necessary)


### AOC update of "Depart Req"
![Updated AOC menu](https://user-images.githubusercontent.com/64070348/150653431-ab683247-379e-4614-83d0-758b455d1d6e.png)

### PDC request
![grafik](https://user-images.githubusercontent.com/64070348/150653495-a9b63042-96c5-4150-812b-0836435fc33d.png)

### ATC Menu (MSN7560)
![grafik](https://user-images.githubusercontent.com/64070348/150653511-96f7eb27-d339-422a-982f-16301d661fba.png)

### ATC Notification
![grafik](https://user-images.githubusercontent.com/64070348/150653521-d31e4560-d310-407b-90cc-f6c876558347.png)
![grafik](https://user-images.githubusercontent.com/64070348/150653538-fe29f98a-17d2-42b9-9994-58a971f57331.png)

### Connection status
![grafik](https://user-images.githubusercontent.com/64070348/150653549-38eb63ff-46b6-4ba4-a71a-023eee5a3335.png)
![grafik](https://user-images.githubusercontent.com/64070348/150653565-476a23e4-88de-4eaf-8e43-692051a860dc.png)

### MSG Record
![grafik](https://user-images.githubusercontent.com/64070348/150653581-67cc9055-cea5-49a3-bfbb-92e6433be2a5.png)

### ATC MSG Display
![grafik](https://user-images.githubusercontent.com/64070348/150653589-72b239f9-ebfc-4e77-8284-c5cc5e041dc4.png)

### Full ATC command process on DCDU
![grafik](https://user-images.githubusercontent.com/64070348/150653612-e8b3ed89-f702-4359-a5ef-bbe67bbef380.png)
![grafik](https://user-images.githubusercontent.com/64070348/150653623-7a40fcd3-6bd0-45e8-8c2e-882a943dac96.png)
![grafik](https://user-images.githubusercontent.com/64070348/150653634-cbe9c3a6-8057-4e1f-8d71-110da2fac783.png)

### ATC MSG button
![grafik](https://user-images.githubusercontent.com/64070348/150654050-addd0c2d-c9d2-4954-9891-83e87aac76c0.png)

### EFB ATSU/AOC
![grafik](https://user-images.githubusercontent.com/64070348/150653970-531340a6-223c-4ce9-b09b-a7015047ba3d.png)

### EFB Realism
![grafik](https://user-images.githubusercontent.com/64070348/150653998-97313f29-a0b5-4524-959f-0c197f5c636c.png)

## References

A320 FCOM documentation and used "MSN 7560" as a reference.
https://www.youtube.com/watch?v=KOdHYtF28Ts




## Additional context

The PR implements the http-based-communication protocol of hoppie.nl.
The API documentation can be found here:
http://www.hoppie.nl/acars/system/tech.html

After a short conversation with Jeroen (hoppie.nl) about missing CORS-headers is it required that the system uses a proxy to extend the HTTP headers of the server's response. This can be done by 'cors-anywhere'.


Discord username (if different from GitHub): sven

## Testing instructions


- Create a Hoppie-ID (http://www.hoppie.nl/acars/system/register.html)
- Activate before every flight Hoppe in the EFB (Settings/ATSU)
- 1. option: Fly online on VATSIM or IVAO and login to online stations (ATC/Communication/Notification)
- 2. option: Download http://www.hoppie.nl/acars/prg/atc/ and be your own controller
   - Start ACARS ATC
   - Logon as a station (i.e. TEST)
   - Logon on the station (ATC/Communication/Notification)

- Links to check if the communication is working well
   - List of all online stations and aircrafts: http://www.hoppie.nl/acars/system/online.html
   - List of all sent and received messages for a specific callsign: http://www.hoppie.nl/acars/system/callsign.html?network=IVAO&callsign=CALLSIGN
   
Also check the initial draft documentation:
https://docs-git-fork-frankkopp-featguidehoppie-flybywire.vercel.app/fbw-a32nx/feature-guides/hoppie/


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
